cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaSpecChecker.java

```diff
@@ -0,0 +1,139 @@
+     *                     this class to include awareness of what defaults are applied.
+     * @param zkCluster The model generated based on the spec. This is requested so that default
+     *                     values not included in the spec can be taken into account, without needing
+     *                     this class to include awareness of what defaults are applied.
+     */
+    public KafkaSpecChecker(KafkaSpec spec, KafkaCluster kafkaCluster, ZookeeperCluster zkCluster) {
+        this.spec = spec;
+        this.kafkaCluster = kafkaCluster;
+        this.zkCluster = zkCluster;
+        this.timestamp = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ").format(new Date());
+    }
+
+
+    public List<Condition> run() {
+        List<Condition> notifications = new ArrayList<>();
+        checkKafkaLogMessageFormatVersion(notifications);
+        checkKafkaStorage(notifications);
+        checkZooKeeperStorage(notifications);
+        checkZooKeeperReplicas(notifications);
+        return notifications;
+    }
+
+    private final static Pattern VERSION_REGEX = Pattern.compile("(\\d\\.\\d+).*");
+
+    /**
+     * Checks if the version of the Kafka brokers matches any custom log.message.format.version config.
+     *
+     * Updating this is the final step in upgrading Kafka version, so if this doesn't match it is possibly an
+     * indication that a user has updated their Kafka cluster and is unaware that they also should update
+     * their format version to match.
+     *
+     * @param warnings List to add a warning to, if appropriate.
+     */
+    private void checkKafkaLogMessageFormatVersion(List<Condition> warnings) {
+        String logMsgFormatVersion = kafkaCluster.getConfiguration().getConfigOption(KafkaConfiguration.LOG_MESSAGE_FORMAT_VERSION);
+        String kafkaBrokerVersion = spec.getKafka().getVersion();
+        if (logMsgFormatVersion != null && kafkaBrokerVersion != null) {
+            Matcher m = VERSION_REGEX.matcher(logMsgFormatVersion);
+            if (m.find() && !kafkaBrokerVersion.startsWith(m.group(0))) {
+                warnings.add(buildCondition("KafkaLogMessageFormatVersion",
+                                            "log.message.format.version does not match the Kafka cluster version, which suggests that an upgrade is incomplete."));
+            }
+        }
+    }
+
+    /**
+     * Checks for a single-broker Kafka cluster using ephemeral storage. This is potentially a problem as it
+     * means any restarts of the broker will result in data loss, as the single broker won't allow for any
+     * topic replicas.
+     *
+     * @param warnings List to add a warning to, if appropriate.
+     */
+    private void checkKafkaStorage(List<Condition> warnings) {
+        if (kafkaCluster.getReplicas() == 1 &&
+            kafkaCluster.getStorage() != null && Storage.TYPE_EPHEMERAL.equals(kafkaCluster.getStorage().getType())) {
```

Comment:
IIRC it's possible to have JBOD storage with ephemeral volumes, so I think you need to cover that case too.

Reply:
Yeah that's right.